research/delf/delf/python/training/global_features/train.py

```diff
@@ -0,0 +1,719 @@
+          decay_rate=exp_decay)
+
+  # Define the optimizer.
+  if FLAGS.optimizer == 'sgd':
+    opt = tfa.optimizers.extend_with_decoupled_weight_decay(
+            tf.keras.optimizers.SGD)
+    optimizer = opt(weight_decay=FLAGS.weight_decay,
+                    learning_rate=lr_scheduler, momentum=FLAGS.momentum)
+  elif FLAGS.optimizer == 'adam':
+    opt = tfa.optimizers.extend_with_decoupled_weight_decay(
+            tf.keras.optimizers.Adam)
+    optimizer = opt(weight_decay=FLAGS.weight_decay,
+                    learning_rate=lr_scheduler)
+
+  # Initializing logging.
+  writer = tf.summary.create_file_writer(model_directory)
+  tf.summary.experimental.set_step(1)
+
+  # Setting up the checkpoint manager.
+  checkpoint = tf.train.Checkpoint(optimizer=optimizer, model=model)
+  manager = tf.train.CheckpointManager(
+          checkpoint,
+          model_directory,
+          max_to_keep=10,
+          keep_checkpoint_every_n_hours=3)
+  if FLAGS.resume:
+    # Restores the checkpoint, if existing.
+    global_features_utils.debug_and_log('>> Continuing from a checkpoint.')
+    checkpoint.restore(manager.latest_checkpoint)
+
+  # Launching tensorboard if required.
+  if FLAGS.launch_tensorboard:
+    tensorboard = tf.keras.callbacks.TensorBoard(model_directory)
+    tensorboard.set_model(model=model)
+    tensorboard_utils.launch_tensorboard(log_dir=model_directory)
+
+  # Log flags used.
+  global_features_utils.debug_and_log('>> Running training script with:')
+  global_features_utils.debug_and_log('>> logdir = {}'.format(model_directory))
+
+  if FLAGS.training_dataset.startswith('retrieval-SfM-120k'):
+    train_dataset = sfm120k.CreateDataset(
+            data_root=FLAGS.data_root,
+            mode='train',
+            imsize=FLAGS.image_size,
+            nnum=FLAGS.neg_num,
+            qsize=FLAGS.query_size,
+            poolsize=FLAGS.pool_size
+    )
+    if FLAGS.val:
+      val_dataset = sfm120k.CreateDataset(
+              data_root=FLAGS.data_root,
+              mode='val',
+              imsize=FLAGS.image_size,
+              nnum=FLAGS.neg_num,
+              qsize=float('Inf'),
+              poolsize=float('Inf'),
+              eccv2020=FLAGS.val_eccv2020
+      )
+  else:
+    train_dataset = tuples_dataset.TuplesDataset(
+            name=FLAGS.training_dataset,
+            data_root=FLAGS.data_root,
+            mode='train',
+            imsize=FLAGS.image_size,
+            nnum=FLAGS.neg_num,
+            qsize=FLAGS.query_size,
+            poolsize=FLAGS.pool_size
+    )
+    if FLAGS.val:
+      val_dataset = tuples_dataset.TuplesDataset(
+              name=FLAGS.training_dataset,
+              data_root=FLAGS.data_root,
+              mode='val',
+              imsize=FLAGS.image_size,
+              nnum=FLAGS.neg_num,
+              qsize=float('Inf'),
+              poolsize=float('Inf')
+      )
+
+  output_types = [tf.float32 for i in range(2 + FLAGS.neg_num)]
```

Comment:
"output_types" -- the variable naming could be improved... "output_types" of what?